File: harness/src/harness.ts
/**
 * Shared model and paths of the subagent harness.
 *
 * The state lives on disk (`features.json`, `progress/`, `specs/`), not in chat.
 */

import { spawnSync } from "node:child_process";
import { existsSync, statSync } from "node:fs";
import { dirname, join, relative, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const HERE = dirname(fileURLToPath(import.meta.url));

/** The checkout this copy runs from: harness/src/ → harness/ → checkout. */
export const CHECKOUT = resolve(HERE, "..", "..");

/**
 * The checkout that owns the harness state.
 *
 * A worktree is a second checkout of the same repository, and the harness is
 * one state machine per *repository*: two features running side by side in two
 * worktrees still share one `features.json` and one id sequence. Forge resolves
 * the very same thing daemon-side from the project root (ADR-012), so the two
 * ends have to agree or the GUI reads a different file than the agent writes.
 *
 * `git rev-parse --git-common-dir` is that agreement — from any worktree it
 * answers the *main* checkout's `.git`. `FORGE_HARNESS_ROOT`, which the daemon
 * exports into every session it starts, wins when it is set, and this checkout
 * is the fallback for a copy that is not in a git repository at all.
 */
function harnessRoot(): string {
  const forced = process.env.FORGE_HARNESS_ROOT?.trim();
  if (forced) return resolve(forced);
  const probe = spawnSync("git", ["rev-parse", "--path-format=absolute", "--git-common-dir"], {
    cwd: CHECKOUT,
    encoding: "utf8",
  });
  const common = probe.status === 0 ? probe.stdout.trim() : "";
  // `<main>/.git` → `<main>`; a bare or unusual layout is left alone.
  return common.endsWith("/.git") ? dirname(common) : CHECKOUT;
}

/** Repo root that holds `harness/`, shared by every worktree of the project. */
export const ROOT = harnessRoot();
export const HARNESS = join(ROOT, "harness");
export const FEATURES = join(HARNESS, "features.json");
export const PROGRESS = join(HARNESS, "progress");
export const SPECS = join(HARNESS, "specs");

/**
 * The "what am I doing right now" note of one feature.
 *
 * Per feature rather than one shared `current.md`, because two features can be
 * in flight at once — one per checkout — and they would otherwise overwrite
 * each other's note. The shared file is still read when a feature has no note
 * of its own, so state written before the split is not lost.
 */
export function currentPath(id: number): string {
  const perFeature = join(PROGRESS, `current_${id}.md`);
  if (existsSync(perFeature)) return perFeature;
  const shared = join(PROGRESS, "current.md");
  return existsSync(shared) ? shared : perFeature;
}

export type Status =
  | "pending"
  | "spec_ready"
  | "in_progress"
  | "in_review"
  | "done"
  | "blocked";

export interface Feature {
  id: number;
  slug: string;
  title?: string;
  spec_raw?: string;
  crates?: string[];
  acceptance?: string[];
  status: Status;
  review_rounds?: number;
  /** Failed `scripts/dev check` attempts during implementation (factor 9). */
  gate_attempts?: number;
  /** GitHub issue number when registered via `from-issue`. */
  source_issue?: number;
  created_at?: string;
  /**
   * The checkout the feature is being implemented in, as Forge knows it.
   * Absent on features registered from the CLI, which name no checkout.
   */
  workspace_id?: string;
  /** That checkout's path, for reading this file by hand. */
  workspace_path?: string;
  /** Bumped by every daemon-side transition; a stale one makes a decision a no-op. */
  revision?: number;
  /** Every run of every step, in order. Written by the daemon's transition table. */
  attempts?: Attempt[];
  /** Why the feature is blocked, mirroring the last `feature_blocked` event. */
  blocked_reason?: string;
  [key: string]: unknown;
}

/**
 * One run of one step — the harness's Dispatch row.
 *
 * A retry is a *new* attempt and not a mutation of the previous one, so the
 * trail reads as "the implementer was started three times" rather than as "the
 * implementer is on attempt 3". An attempt with no `settled_at` and no live job
 * is a step that died with its daemon, which is exactly what makes restart
 * recovery possible.
 */
export interface Attempt {
  step: "Spec" | "Implement" | "Review";
  n: number;
  job?: string;
  provider?: string;
  transport?: string;
  started_at: string;
  settled_at?: string;
  outcome?: "succeeded" | "failed" | "cancelled" | "blocked" | "superseded";
  detail?: string;
}

export interface Rules {
  one_feature_at_a_time?: boolean;
  require_human_spec_approval?: boolean;
  require_green_gate_to_close?: boolean;
  max_review_rounds?: number;
  /** Auto-block the implementer after this many gate failures. */
  max_gate_attempts?: number;
  /** How many times the daemon restarts a step whose job failed, before blocking. */
  max_step_attempts?: number;
  /** Wall clock for one attempt; past it the daemon cancels the job. */
  max_step_minutes?: number;
  /** Silence after which an attempt is reported stale — a warning, never a kill. */
  stale_after_minutes?: number;
  /** Require `progress/gate_<id>.md` from `spec_ready` on. */
  require_gate_file?: boolean;
  /** Require `progress/context_<id>.md` from `in_progress` on. */
  require_context_bundle?: boolean;
  valid_status?: string[];
}

export interface HarnessData {
  project?: string;
  description?: string;
  rules?: Rules;
  features?: Feature[];
}

/** Statuses in which the spec must already exist on disk. */
export const SPEC_REQUIRED: ReadonlySet<string> = new Set([
  "spec_ready",
  "in_progress",
  "in_review",
  "done",
]);

/** Statuses that occupy a checkout's active slot. */
export const ACTIVE: ReadonlySet<string> = new Set(["in_progress", "in_review"]);

/**
 * The slot a feature occupies: its checkout, or the shared "no checkout" slot
 * for features registered outside Forge. One active feature per slot — which
 * is the whole of `one_feature_at_a_time` now that a project can be checked
 * out into several worktrees at once.
 */
export function slotOf(f: Feature): string {
  return f.workspace_id ?? "(no checkout)";
}

export const SPEC_FILES = ["requirements.md", "design.md", "tasks.md"] as const;

export const SLUG_RE = /^[a-z0-9]+(-[a-z0-9]+)*$/;

/** Reads and parses `features.json`. Throws if it does not exist or is not valid JSON. */
export async function load(): Promise<HarnessData> {
  const text = await Bun.file(FEATURES).text();
  return JSON.parse(text) as HarnessData;
}

export function features(data: HarnessData): Feature[] {
  return data.features ?? [];
}

export function specDir(f: Pick<Feature, "id" | "slug">): string {
  return join(SPECS, `${f.id}-${f.slug}`);
}

export function progressFile(name: string, id: number): string {
  return join(PROGRESS, `${name}_${id}.md`);
}

export function isDir(path: string): boolean {
  try {
    return statSync(path).isDirectory();
  } catch {
    return false;
  }
}

export function isFile(path: string): boolean {
  try {
    return statSync(path).isFile();
  } catch {
    return false;
  }
}

export { existsSync };

/** Path relative to the repo root, with POSIX separators. */
export function rel(path: string): string {
  return relative(ROOT, path);
}

/** Padding by code points (not by UTF-16 units), like Python. */
function width(s: string): number {
  return [...s].length;
}

export function padEnd(s: string, n: number): string {
  const pad = n - width(s);
  return pad > 0 ? s + " ".repeat(pad) : s;
}

export function padStart(s: string, n: number): string {
  const pad = n - width(s);
  return pad > 0 ? " ".repeat(pad) + s : s;
}

/** What a review file said, once. */
export type Verdict = "APPROVED" | "CHANGES_REQUESTED" | "MISSING" | "AMBIGUOUS";

/**
 * Read the verdict out of a review file.
 *
 * A *line*, never a substring of the document. The rule here used to be
 * `text.includes("APPROVED")`, which passed on the word wherever it appeared —
 * including in `.claude/agents/reviewer.md`'s own format example, which prints
 * `**Verdict:** APPROVED | CHANGES_REQUESTED`, and in the ordinary prose of a
 * rejection ("cannot be approved").
 *
 * Kept in step with `parse_verdict` in `crates/daemon/src/harness_runner.rs`:
 * the daemon decides with it and this validates with it, so they have to agree
 * on what a review file says.
 */
export function parseVerdict(review: string): Verdict {
  let seen: Verdict | null = null;
  for (const line of review.split("\n")) {
    const verdict = verdictOnLine(line);
    if (!verdict) continue;
    if (seen === null) seen = verdict;
    // A reviewer that wrote both answered neither.
    else if (seen !== verdict) return "AMBIGUOUS";
  }
  return seen ?? "MISSING";
}

/**
 * The verdict a single line states, or `null` when it states none.
 *
 * Tolerates the markdown a reviewer actually writes — a bullet, a heading, an
 * emphasised `**Verdict:**` label — and then demands that what is left be
 * *exactly* one token. That last part is what makes the unfilled
 * `APPROVED | CHANGES_REQUESTED` template line match nothing.
 */
function verdictOnLine(line: string): "APPROVED" | "CHANGES_REQUESTED" | null {
  let text = line.trim().replace(/^[-*#>\s]+/, "");
  text = text.replace(/^[*_]+/, "").trim();
  const labelled = /^\**\s*verdict\s*\**\s*:(.*)$/i.exec(text);
  if (labelled) text = labelled[1];
  text = text.trim().replace(/^[*_`\s]+|[*_`\s]+$/g, "").trim();
  const token = text.toUpperCase();
  if (token === "APPROVED") return "APPROVED";
  // `REJECTED` is accepted beside `CHANGES_REQUESTED` because review files
  // written by earlier builds say it.
  if (token === "CHANGES_REQUESTED" || token === "REJECTED") return "CHANGES_REQUESTED";
  return null;
}

File: crates/domain/src/harness.rs
//! Harness feature state mirrored from `harness/features.json` (runtime-only on
//! the wire — no SQLite column).

use crate::ids::{SessionId, WorkspaceId};
use serde::{Deserialize, Serialize};

/// One row from `harness/features.json`.
#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct HarnessFeature {
    pub id: u32,
    pub slug: String,
    #[serde(default)]
    pub title: Option<String>,
    #[serde(default)]
    pub spec_raw: Option<String>,
    #[serde(default)]
    pub status: String,
    #[serde(default)]
    pub review_rounds: Option<u32>,
    #[serde(default)]
    pub gate_attempts: Option<u32>,
    #[serde(default)]
    pub crates: Option<Vec<String>>,
    #[serde(default)]
    pub acceptance: Option<Vec<String>>,
    #[serde(default)]
    pub source_issue: Option<u32>,
    #[serde(default)]
    pub created_at: Option<String>,
    /// Live orchestrator session, set by the GUI when the feature starts.
    #[serde(default)]
    pub orchestrator_session_id: Option<SessionId>,
    /// The checkout this feature is being implemented in.
    ///
    /// The state itself is one file per *repository* (see
    /// `harness_service`), so two features can run side by side in two
    /// worktrees of the same project; this is what tells them apart, and what
    /// the diff and commit actions of the Feature tab are aimed at. `None` on
    /// rows registered outside Forge, where there is no checkout to name.
    #[serde(default)]
    pub workspace_id: Option<WorkspaceId>,
    /// The same checkout as a path, for whoever is reading `features.json`
    /// by hand — a `WorkspaceId` means nothing outside the daemon's model.
    #[serde(default)]
    pub workspace_path: Option<String>,
    /// How many times this row has been written, so a decision taken against
    /// a stale view of it is refused instead of applied twice.
    ///
    /// A client reads a feature, a person clicks *Approve*, and in between the
    /// review that was already running settled the row. Replaying that click
    /// used to append a second `human_gate_resolved` and launch a second
    /// implementer in the same worktree. The client sends the revision it saw;
    /// a mismatch is a no-op that answers with the current row.
    #[serde(default)]
    pub revision: Option<u64>,
    /// Every run of a step, in order — the harness's Dispatch row.
    ///
    /// `review_rounds` and `gate_attempts` are counters and say nothing about
    /// *which* provider produced a result, how long it took, or whether the
    /// job that was started ever settled. That last one is what a daemon
    /// restart needs: an attempt with no `settled_at` and no live job is a step
    /// that died with its daemon, and it is recoverable precisely because it is
    /// written down.
    #[serde(default)]
    pub attempts: Option<Vec<HarnessAttempt>>,
    /// Why the feature is `blocked`, in one sentence.
    ///
    /// Also the last `feature_blocked` event's `reason`, but a client showing
    /// a blocked card should not have to read the timeline to render it.
    #[serde(default)]
    pub blocked_reason: Option<String>,
}

/// One run of one step: who ran it, when, and how it ended.
///
/// Orca's Dispatch at feature granularity. A retry is a *new* attempt rather
/// than a mutation of the previous one, so the trail reads as "the implementer
/// was started three times" and not as "the implementer is on attempt 3".
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct HarnessAttempt {
    pub step: HarnessStep,
    /// 1-based, counted per step and reset when the step is left behind.
    pub n: u32,
    /// The headless run, when this attempt was one.
    #[serde(default)]
    pub job: Option<String>,
    #[serde(default)]
    pub provider: Option<String>,
    /// `cli` today; `acp` once the second transport lands.
    #[serde(default)]
    pub transport: Option<String>,
    pub started_at: String,
    #[serde(default)]
    pub settled_at: Option<String>,
    /// `succeeded` | `failed` | `cancelled` | `blocked`, unset while it runs.
    #[serde(default)]
    pub outcome: Option<String>,
    /// Why it ended that way, when that needs saying.
    #[serde(default)]
    pub detail: Option<String>,
    /// The provider's own session id, when the stream named one.
    ///
    /// What a retry passes as `resume_from` so the next attempt re-enters the
    /// same conversation instead of throwing away research already paid for.
    #[serde(default)]
    pub provider_session_id: Option<String>,
}

impl HarnessAttempt {
    /// Whether this attempt is still on record as running.
    #[must_use]
    pub fn is_live(&self) -> bool {
        self.settled_at.is_none()
    }
}

impl HarnessFeature {
    /// Whether the feature still occupies its checkout: anything that is not
    /// finished or abandoned. A second feature may not start in a checkout
    /// that already has one of these.
    #[must_use]
    pub fn is_open(&self) -> bool {
        !matches!(self.status.as_str(), "done" | "blocked")
    }
}

/// Parsed `harness/features.json`.
#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct HarnessFeatureList {
    #[serde(default)]
    pub project: Option<String>,
    #[serde(default)]
    pub features: Vec<HarnessFeature>,
    #[serde(default)]
    pub initialized: bool,
}

/// One line from `harness/progress/events_<id>.jsonl`.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct HarnessEvent {
    pub ts: String,
    pub kind: String,
    /// Remaining fields from the JSON line, serialized for display.
    #[serde(default)]
    pub detail: String,
    /// The headless run this event started, when it started one.
    ///
    /// Lifted out of `detail` because it is the one field a client *acts* on
    /// rather than prints: jobs live in the daemon's memory, so a restart
    /// forgets every row while the transcripts stay on disk. This id is what
    /// still points at one, and reading it here keeps JSON parsing in the
    /// crate that already does it.
    #[serde(default)]
    pub job: Option<crate::ids::JobId>,
}

/// Which markdown artefact to read under `harness/progress/` or `harness/specs/`.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[non_exhaustive]
pub enum HarnessArtifactKind {
    Gate,
    Context,
    Impl,
    Review,
    Requirements,
    Design,
    Tasks,
    Current,
    #[serde(other)]
    Unknown,
}

/// Human or UI-driven step in the harness cycle.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[non_exhaustive]
pub enum HarnessAdvanceAction {
    ApproveSpec,
    ReviseSpec,
    Block {
        reason: String,
    },
    StartImplement,
    StartReview,
    /// Run the step the feature was blocked on again, from `blocked`.
    ///
    /// The counterpart of the retry budget: once the machine has spent it, the
    /// only thing that may start another attempt is a person who has looked at
    /// why it failed.
    RetryStep,
    /// Take a `done` or `blocked` feature back to `pending`.
    Reopen,
}

/// One step of the harness cycle, as a thing that can be *run*.
///
/// The names are the roles the harness has always had; what is new is that
/// each is a job with an exit code rather than an agent somebody watches. The
/// human gate is deliberately not here: approving a spec is not a step that
/// can be run, it is the one moment the machine stops and waits.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[non_exhaustive]
pub enum HarnessStep {
    /// Research the codebase and write the spec, then stop at the gate.
    Spec,
    /// Implement the approved spec and run the gate command.
    Implement,
    /// Judge the implementation and return a verdict.
    Review,
}

impl HarnessStep {
    /// The status the feature takes while this step runs.
    #[must_use]
    pub fn running_status(self) -> &'static str {
        match self {
            Self::Spec => "pending",
            Self::Implement => "in_progress",
            Self::Review => "in_review",
        }
    }

    /// The event name appended to the feature's log when it starts.
    #[must_use]
    pub fn started_event(self) -> &'static str {
        match self {
            Self::Spec => "spec_started",
            Self::Implement => "impl_started",
            Self::Review => "review_started",
        }
    }
}

impl HarnessFeature {
    #[must_use]
    pub fn is_active(&self) -> bool {
        matches!(self.status.as_str(), "in_progress" | "in_review")
    }

    #[must_use]
    pub fn needs_gate(&self) -> bool {
        self.status == "spec_ready"
    }
}
